pkg/cache/v3/linear.go

```diff
@@ -21,46 +21,73 @@ import (
 	"strconv"
 	"strings"
 	"sync"
-	"sync/atomic"
 
 	"github.com/envoyproxy/go-control-plane/pkg/cache/types"
 	"github.com/envoyproxy/go-control-plane/pkg/log"
 )
 
+// cachedResource is used to track resources added by the user in the cache.
+// It contains the resource itself and its associated version (currently in two different modes).
+// TODO(valerian-roche): store serialized resource if already marshaled to compute the version.
+type cachedResource struct {
+	types.Resource
+
+	// cacheVersion is the version of the cache at the time of last update, used in sotw.
+	cacheVersion string
+	// resourceVersion is the version of the resource itself (built through stable marshaling).
+	// It is only set if computeResourceVersion is set to true on the cache.
+	resourceVersion string
+}
+
+type watches struct {
+	// sotw keeps track of current sotw watches, indexed per watch id.
+	sotw map[uint64]ResponseWatch
```

Comment:
A potential evolution is to make this an interface with a "compute response" and a "send response" methods that would allow to fully remove the delta vs. sotw difference here.